Discordに投稿するさい、タイトルと共にその項目の詳細を表示する

esehara opened the pull request:
# 動作

![image](https://user-images.githubusercontent.com/615042/229334656-33de3862-8577-4f8c-8887-aba0a5cf3a10.png)

# 目的

現状の人工発達Botでは、wikipediaの項目と共に、その項目の詳細を別途表示させています。これによって: 

 - 項目を開かなくてもその項目について知ることができる
 - 項目に詳細が無い場合でも、ある程度のその項目の情報を得ることができる
   - たまにDiscordのURL展開機能で開いても、その先に詳細のない項目がある

というメリットがあります。また、wikipediaの項目によってはnsfw (職場閲覧注意）の画像を利用した項目も存在していますので、できるだけテキストで完結させて、うっかりURL展開でそのような画像を表示するのを防止する意味合いもあります。
 
# 以下、このpull requestの詳細

## `-t` のオプションフラグを作成

これはローカルでテストをするさいに、Discordを投稿しないでコンソールに文面を表示するためのオプションです。これにより、`BOT TOKEN`を所持していなくても、文面自体の改善は可能になります。

## 重複部分の共通化

まず、wikipediaのランダムAPIについては、項目の詳細については習得できません。従って、別途で項目の詳細を習得してくる必要があります。これはwikipediaの検索APIで、ランダムAPIから取得した単語を検索するという方法で解決しています。

この場合、ランダムAPIと検索APIで共通の処理が出てきます（共通のパラメータをセットする部分と、APIを叩く部分）。前者については少しやりすぎかもしれません。

## 項目詳細部分についてのタグの排除

wikipediaの項目詳細の部分を引っ張ると、どうやらキーワードなどにタグが含まれていることがあるので、それも別途処理してテキストだけにするための再帰的な関数(`getTextFromHTML`関数がそれにあたる）も実装しています。

# 気になる部分

単純に二回WikipediaのAPIを叩くので、直感的にスマートではない感じがするのと、通信部分で二回分、通信の不具合などでリクエストが通らないなどのリスクがあります。あとの判断はお任せします。

# 本音を言うと

単にgo言語をいじってみたかったのでそれだけです。

esehara committed:
テスト用のオプションを作成する

Botを外部から開発する場合、トークンが無いと文章がチェックできないのは
大変なので、Botが投稿するであろう文章を生成するだけのflagを作る。

esehara committed:
コマンドで文章を生成する部分を関数として切り離す

esehara committed:
Wikipediaから項目の詳細を取ってきて、テキストに表示する

esehara committed:
go fmtを走らせる

esehara committed:
URLを展開しないように`<>`でURLを囲む

esehara committed:
コマンドの文章がちょっとおかしいので変更

esehara committed:
Merge branch 'main' of https://github.com/tenax66/serendip into show-detail

esehara committed:
検索結果を出来るだけ絞り込む

rnlimitのようなパロメータが検索にもあるので、それを使って検索結果を絞り込む

esehara committed:
基礎的な部分で間違いがあったので、細かい修正を行う

tenax66 commented on main.go:
```diff
@@ -61,17 +74,67 @@ func main() {
+	resp, err := getWikipediaAPI(params)
+	defer resp.Body.Close()
+
+	var result WikipediaSearchResult
```
必要がない限りvarキーワードは使わない（`:=`を用いる）方針としたいので、書き換えていただきたいです。

tenax66 commented on main.go:
```diff
@@ -61,17 +74,67 @@ func main() {
+	}
+	var content string
+	if len(result.Query.Search) == 0 {
+		content = pageTitle + "\n" + "<" + pageURL + ">"
```
```suggestion
	if len(result.Query.Search) == 0 {
		content := pageTitle + "\n" + "<" + pageURL + ">"
```
同上で、ここの`var`宣言は冗長なので`:=`に置き換えてほしいです。

tenax66 commented on main.go:
```diff
@@ -2,20 +2,33 @@ package main
-	"github.com/bwmarrin/discordgo"
+var (
+	testOpt = flag.Bool("t", false, "テスト用です。投稿する文章をDiscordを経由せず表示します")
 )
```
```suggestion
var testOpt = flag.Bool("t", false, "テスト用です。投稿する文章をDiscordを経由せず表示します")
```
Goに詳しくないのでふつうこういう書き方をするものか知らないのですが、カッコは不要であれば外しておいてください。

tenax66 commented on main.go:
```diff
@@ -61,17 +74,67 @@ func main() {
+
+		pageDetail := getTextFromHTML(rawSnippet)
+
+		content = pageTitle + "\n\n" + pageDetail + "\n\n" + "<" + pageURL + ">"
```
```suggestion
		content := pageTitle + "\n\n" + pageDetail + "\n\n" + "<" + pageURL + ">"
```
上の変数宣言を削ることにより、こちらも直すことになります。

tenax66 commented on main.go:
```diff
@@ -84,22 +147,20 @@ func onSlashCommand(s *discordgo.Session, i *discordgo.InteractionCreate) {
 
 	// レスポンスをパース
-	var result WikipediaAPIResult
+	var result WikipediaRandomResult
```
同上です。（なるべく`:=` を用いる方針）

esehara replied:
これは[元々のソースに宣言が入っていて](https://github.com/tenax66/serendip/blob/main/main.go#L102)、それを踏襲しています。(`json.NewDecoder(resp.Body).Decode(&result)`で使われている奴ですね）

tenax66 replied:
失礼しました、了解です。

tenax66 commented on main.go:
```diff
@@ -113,13 +174,38 @@ func getRandomPage() (string, string, error) {
+		return ""
+	}
+
+	var text string
+	for c := n.FirstChild; c != nil; c = c.NextSibling {
+		text += getTextFromHTML(c)
+	}
```
`var`宣言を使わない方針は同様です。
文字列連結には`+=`ではなく`strings.Builder`等を利用してほしいです。

tenax66 commented on main.go:
```diff
@@ -61,17 +74,67 @@ func main() {
+	return resp, err
+}
+
+func getDiscordMessage() (string, error) {
```
`getDiscordMessage`メソッド全体についてのreviewです。
要約を取得するための処理の分岐が複雑に見えるので、`getTextFromHTML`メソッド内に処理を移して、このメソッドの見た目をすっきりさせてほしいです。

`getDiscordMessage`は`getRandomPage`と`getTextFromHTML`の結果を結合して実際にテキストチャンネルに送信される文字列を作るメッセージだと読みました。
そうだとすると、このメソッドの処理は単にふたつのメソッドの出力結果を結合するだけ、というようになるのが望ましいと思います。しかし、現状では`getTextFromHTML`に関係する処理の比重が高くて、両メソッドが対等になっていないように見えます。

tenax66 commented on main.go:
```diff
@@ -113,13 +174,38 @@ func getRandomPage() (string, string, error) {
 	return page.Title, "https://ja.wikipedia.org/wiki/" + url.PathEscape(page.Title), nil
 }
 
+func getTextFromHTML(n *html.Node) string {
```
このメソッドはwikipedia記事の一部を取得するためのものだと思うのですが、とすると、`getTextFromHTML`という名前は抽象的に思えます。具体的に何を取得するのかが明確にわかる名前に変更してほしいです。

esehara replied:
これはWikipedia記事の一部を習得したあとに 、検索した文字列にマッチしている部分が`span`タグで囲まれている場合があるので、それを削除するためのものものなのですが([例](https://ja.wikipedia.org/w/api.php?action=query&list=search&srsearch=iphone&format=json&srlimit=1))、もう少しわかりやすいメソッド名を考えてみます。

tenax66 replied:
了解しました、整理ありがとうございます！

tenax66 reviewed (changes requested):
コードレビューを実施しました。

1. `var`宣言はなるべく`:=`で置き換えたい
2. `getDiscordMessage` メソッドで`getRandomPage`と`getTextFromHTML`の結果を結合する処理の見通しが悪いため、処理を`getTextFromHTML`に寄せてほしい

という2点が主な観点となっています。ご確認・修正お願いします！

esehara commented:
ありがとうございます。早速修正します。

esehara committed:
文章を組み立てるための関数を生成する

恐らく事例としては考えにくいが、念のためにWikipediaのページに詳細が見つからない
場合の分岐を作っている。

このときに、出来るだけ`var`の宣言を減らしたいので、ifの内部で代入するよりは、
関数化して、それぞれの分岐から文字列を返し、その結果を変数に代入したほうが、
スッキリするかと思われるのでそのような形にした。

esehara committed:
testOptの部分のカッコを外す

esehara committed:
getDiscordMessegeを整理し、別途にgetSearchPageを分割する

getRandomPage（ランダムに項目を拾ってくるメソッド）は存在するのに、
項目から詳細を拾ってくる処理はベタ書きなのはアンバランスなので、
二つの処理をそれぞれ関数としてわけて、スッキリさせる

esehara committed:
文字列の結合に`+=`を使うのではなく、`strings.Builder`を使う

esehara committed:
関数名をもう少しわかりやすい名前に変更する

この関数の目的は、「検索で引っかかったページの単語の部分が`span`タグで囲まれている。
このタグを削除する」というものなので、そういった処理であることを関数名で
ある程度明示するようにした。

esehara committed:
明らかにいらない分岐処理を削除する

これは`span`タグを消すのが主目的なので、`script`タグは出て来ないと
判断し、この分岐は削除した。

tenax66 reviewed (approved):
確認しました、OKです。細かい点まで修正いただきありがとうございます！
